gym/utils/save_video.py

```diff
@@ -0,0 +1,109 @@
+"""Utility functions to save rendering videos."""
+import os
+from typing import Callable, Optional
+
+import gym
+from gym import logger
+
+try:
+    from moviepy.video.io.ImageSequenceClip import ImageSequenceClip
+except ImportError:
+    raise gym.error.DependencyNotInstalled(
+        "MoviePy is not installed, run `pip install moviepy`"
+    )
+
+
+def capped_cubic_video_schedule(episode_id: int) -> bool:
+    """The default episode trigger.
+
+    This function will trigger recordings at the episode indices 0, 1, 4, 8, 27, ..., :math:`k^3`, ..., 729, 1000, 2000, 3000, ...
+
+    Args:
+        episode_id: The episode number
+
+    Returns:
+        If to apply a video schedule number
+    """
+    if episode_id < 1000:
+        return int(round(episode_id ** (1.0 / 3))) ** 3 == episode_id
+    else:
+        return episode_id % 1000 == 0
+
+
+def save_video(
+    frames: list,
+    video_folder: str,
+    episode_trigger: Callable[[int], bool] = None,
+    step_trigger: Callable[[int], bool] = None,
+    video_length: Optional[int] = None,
+    name_prefix: str = "rl-video",
+    episode_index: int = 0,
+    step_starting_index: int = 0,
+    **kwargs,
+):
+    """Save videos from rendering frames.
+
+    This function extract video from a list of render frame episodes.
+
+    Args:
+        frames (List[RenderFrame]): A list of frames to compose the video.
+        video_folder (str): The folder where the recordings will be stored
+        episode_trigger: Function that accepts an integer and returns ``True`` iff a recording should be started at this episode
+        step_trigger: Function that accepts an integer and returns ``True`` iff a recording should be started at this step
+        video_length (int): The length of recorded episodes. If it isn't specified, the entire episode is recorded.
+            Otherwise, snippets of the specified length are captured.
+        name_prefix (str): Will be prepended to the filename of the recordings.
+        episode_index (int): The index of the current episode.
+        step_starting_index (int): The step index of the first frame.
+        **kwargs: The kwargs that will be passed to moviepy's ImageSequenceClip.
+            You need to specify either fps or duration.
+
+    Example:
+        >>> import gym
+        >>> from gym.utils.save_video import save_video
+        >>> env = gym.make("FrozenLake-v1", render_mode="rgb_array")
+        >>> env.reset()
+        >>> step_starting_index = 0
+        >>> episode_index = 0
+        >>> for step_index in range(199):
+        ...    action = env.action_space.sample()
+        ...    _, _, done, _ = env.step(action)
+        ...    if done:
+        ...       save_video(
+        ...          env.render(),
```

Comment:
It looks like here only calls `env.render()` when the env is done. Does this mean when calling `env = gym.make("FrozenLake-v1", render_mode="rgb_array")` each frame is rendered and cached? If this is the case this could be an undesirable bottleneck — we might only keep the rendered frames of a few episodes.

Reply:
Yes, with the current Render API `env.render()` returns the list of frames of the whole episode when the mode is `rgb_array`; old-like behavior can be obtained with `single_rgb_array`.

When you call `render()` or `reset()`, the frame list is cleaned

Reply:
Sorry, I am struggling to understand this. If I don’t want to render every frame, how can I achieve it with the current API? What is the old-like behavior with `single_rgb_array`?

Reply:
If you want to render a single frame representing the current state, you can achieve it with `single_rgb_array`, thus similar to before:

```python
env = gym.make("MyEnv", render_mode="single_rgb_array")
env.reset()

for _ in range(n_steps):
    env.step(...)
    env.render()  # single frame

env.close()
```